File: 2-rbac-and-guards/backend/0-typescript/src/modules/auth/auth.service.ts
/**
 * Business logic service for Auth.
 */
import {
    ConflictException,
    Injectable,
    UnauthorizedException,
} from "@nestjs/common"
import {
    JwtService,
} from "@nestjs/jwt"
import {
    InjectRepository,
} from "@nestjs/typeorm"
import * as bcrypt from "bcrypt"
import {
    Repository,
} from "typeorm"
import {
    Role,
} from "../../common"
import {
    UserEntity,
} from "../../entities"
import {
    SignInDto,
    SignUpDto,
} from "./dto"

/**
 * Issues JWT embedding role claims for downstream authorization.
 */
@Injectable()
export class AuthService {
    constructor(
    @InjectRepository(UserEntity)
    private readonly usersRepo: Repository<UserEntity>,
    private readonly jwtService: JwtService,
    ) {}

    /**
     * Registers a new user — prevents duplicate emails and stores a bcrypt hash (never plain password).
     * Role defaults to USER unless the caller provides a specific role (demo convenience).
     *
     * @param dto — signup payload containing email, password, and optional role
     * @returns Confirmation message object
     * @throws ConflictException (409) when the email is already taken
     */
    async signUp(dto: SignUpDto) {
        // Guard: reject duplicate emails — uniqueness is enforced here before the DB constraint fires.
        const existing = await this.usersRepo.findOne({
            where: {
                email: dto.email,
            },
        })
        if (existing) {
            throw new ConflictException("Email already registered")
        }

        // Hash with bcrypt cost factor 10 — never store plain-text passwords.
        const hash = await bcrypt.hash(dto.password,
            10)

        // Default to USER role so admin accounts can only be created explicitly.
        const role = dto.role ?? Role.USER

        await this.usersRepo.save(
            this.usersRepo.create({
                email: dto.email,
                password: hash,
                role,
            }),
        )
        return {
            message: "Created",
        }
    }

    /**
     * Validates credentials and issues a signed JWT embedding the user's role.
     * The role inside the token lets the RolesGuard make decisions without
     * a database round-trip on every request.
     *
     * @param dto — sign-in payload containing email and password
     * @returns Object with `access_token` (JWT)
     * @throws UnauthorizedException (401) on wrong email or password
     */
    async signIn(dto: SignInDto) {
        // Look up user by email; bcrypt.compare checks the password hash.
        const user = await this.usersRepo.findOne({
            where: {
                email: dto.email,
            },
        })

        // Reject early if user not found OR password hash mismatch — same error to prevent email enumeration.
        if (!user || !(await bcrypt.compare(dto.password,
            user.password))) {
            throw new UnauthorizedException("Invalid credentials")
        }

        // Embed `sub` (user id) + `role` in the JWT payload so the guard can read the role without DB.
        const access_token = await this.jwtService.signAsync({
            sub: user.id,
            role: user.role,
        })
        return {
            access_token,
        }
    }
}

File: 2-rbac-and-guards/backend/1-java/src/main/java/com/example/rbacdemo/service/AuthService.java
package com.example.rbacdemo.service;

import com.example.rbacdemo.dto.SignInRequest;
import com.example.rbacdemo.dto.SignInResponse;
import com.example.rbacdemo.dto.SignUpRequest;
import com.example.rbacdemo.entity.User;
import com.example.rbacdemo.repository.UserRepository;
import com.example.rbacdemo.security.JwtService;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.HashMap;
import java.util.Map;

/**
 * Business logic for authentication.
 * Handles user registration (signup) and credential validation + JWT issuance (signin).
 */
@Service
public class AuthService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    /**
     * @param userRepository  JPA repository for user persistence
     * @param passwordEncoder BCrypt encoder injected from SecurityConfig
     * @param jwtService      generates and validates JWTs
     */
    public AuthService(UserRepository userRepository, PasswordEncoder passwordEncoder, JwtService jwtService) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
    }

    /**
     * Registers a new user after checking for email uniqueness.
     * Password is hashed with BCrypt before persistence — never stored as plain text.
     *
     * @param request sign-up payload containing email, password, and optional role
     * @return confirmation map {@code {"message": "Created"}}
     * @throws ResponseStatusException (409 Conflict) if the email is already taken
     */
    public Map<String, String> signUp(SignUpRequest request) {
        // Reject duplicate emails before attempting a DB insert.
        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "Email already registered");
        }

        User user = new User();
        user.setEmail(request.getEmail());
        user.setPassword(passwordEncoder.encode(request.getPassword())); // BCrypt hash — never store plain text

        // Default to "user" role when the caller does not specify one.
        String role = request.getRole();
        if (role == null || role.trim().isEmpty()) {
            role = "user";
        }
        user.setRole(role);

        userRepository.save(user);

        Map<String, String> response = new HashMap<>();
        response.put("message", "Created");
        return response;
    }

    /**
     * Validates credentials and issues a signed JWT embedding the user's role claim.
     * The role in the token lets the SecurityFilterChain make RBAC decisions without a DB query per request.
     *
     * @param request sign-in payload containing email and password
     * @return {@link SignInResponse} wrapping the access token
     * @throws ResponseStatusException (401 Unauthorized) on wrong email or mismatched password
     */
    public SignInResponse signIn(SignInRequest request) {
        // Look up user — throw 401 with a generic message to prevent email enumeration.
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials"));

        // BCrypt compare: raw password vs stored hash — same 401 message on mismatch to avoid enumeration.
        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials");
        }

        // Embed user ID (sub) and role in the JWT so downstream guards can read the role without hitting DB.
        String token = jwtService.generateToken(user.getId(), user.getRole());
        return new SignInResponse(token);
    }
}
